### PR TITLE
ENH: support pathlib in ReaderBase

### PR DESCRIPTION
**Pull request by tylerjereddy**
* the creation of a `Universe` object involving GROMACS formats should now accept `pathlib`
objects as requested in gh-2497; resolving that issue entirely is a separate matter, this branch focuses only on the GROMACS format and only if you proceed via the public `Universe` creation API

* probably could make it work in Cython directly, though I'm not convinced we need to do that for this part of the fix at least, and we probably should still support strings as well of course

### CODE REVIEW

**Review by IAlibay — COMMENTED** [review]
Shouldn't this be a ReaderBase thing instead to store self.filename properly?

**Review by richardjgowers — APPROVED** [review]

**Comment by hmacdope** [comment]
> Shouldn't this be a ReaderBase thing instead to store self.filename properly?

Seconding this, can this not be dealt with at `ReaderBase` level to fix it for all readers?

**Comment by tylerjereddy** [comment]
Ok, I moved the patch to `ReaderBase` and added a test for one more format (`DCD`) as a sniff test for generalization.

I think that's probably good enough for now, but note that:

1) This still doesn't cover all readers because `SingleFrameReaderBase` has its own `self.filename` -- I'd be happy to delay patching/testing that for now.
2) The testing could probably be centralized/parametrized somewhere, covering a bunch of formats rather than testing the same pass-through for just two formats. Maybe good enough for now though..

**Review by IAlibay — APPROVED** [review]
Assuming this passes tests I'm happy for this PR to go through. @tylerjereddy could you raise an issue about the missing bits? Maybe we should move stuff to ProtoReader in the future but it seems fine for this current issue?

**Comment by tylerjereddy** [comment]
Eh, the failure is related--this is the problem with inheritance... I'd just as soon scope this back to GROMACS and move on.

I'll see if I can add a shim to fix it though.

**Comment by tylerjereddy** [comment]
Ok, I pushed in a commit to shim around our `NamedStream` stuff in the base class, and this seems to allow the full suite to pass for me locally at least.

gh-3937 contains the the follow-up reminders I noted above.

**Comment by hmacdope** [comment]
LGTM @tylerjereddy

**Review by hmacdope — APPROVED** [review]

**Comment by tylerjereddy** [comment]
Ok, I have approvals from 3 core developers and CI is passing, in it goes.

**Pull request merged**
